velox/exec/tests/ArrowStreamTest.cpp

```diff
@@ -0,0 +1,250 @@
+          privateData()->lastError_ = "Failed.";
+      }
+      return code;
+    }
+
+    std::shared_ptr<memory::MemoryPool> pool_;
+    struct ArrowArrayStream* stream_;
+  };
+
+  void exportArrowStream(
+      std::shared_ptr<ArrowReader> reader,
+      struct ArrowArrayStream* out) {
+    out->get_schema = ExportedArrayStream::staticGetSchema;
+    out->get_next = ExportedArrayStream::staticGetNext;
+    out->get_last_error = ExportedArrayStream::staticGetLastError;
+    out->release = ExportedArrayStream::staticRelease;
+    out->private_data = new ExportedArrayStream::PrivateData{std::move(reader)};
+  }
+};
+
+TEST_F(ArrowStreamTest, basic) {
+  vector_size_t size = 1'000;
+  std::vector<RowVectorPtr> vectors;
+  for (int32_t i = 0; i < 3; ++i) {
+    vectors.push_back(makeRowVector(
+        {makeFlatVector<bool>(
+             size, [](auto row) { return (row / 10) % 2 == 0; }, nullEvery(3)),
+         makeFlatVector<int8_t>(
+             size, [](auto row) { return row % 100; }, nullEvery(5)),
+         makeFlatVector<int16_t>(
+             size, [&](auto row) { return row * i; }, nullEvery(2)),
+         makeFlatVector<int32_t>(
+             size, [](auto row) { return row; }, nullEvery(5)),
+         makeFlatVector<int64_t>(
+             size,
+             [&](vector_size_t row) { return size * i + row; },
+             nullEvery(5)),
+         makeFlatVector<float>(
+             size, [](vector_size_t row) { return row * 1.1; }, nullEvery(5)),
+         makeFlatVector<double>(
+             size, [](vector_size_t row) { return row * 1.3; }, nullEvery(11)),
+         makeFlatVector<StringView>(
+             size,
+             [](vector_size_t row) {
+               uint64_t val = 100000000000 + (uint64_t)(row % 100);
+               return StringView(std::to_string(val));
+             },
+             nullEvery(7))}));
+  }
+  createDuckDbTable(vectors);
+  auto type = asRowType(vectors[0]->type());
+  auto reader = std::make_shared<ArrowReader>(pool_, vectors, type);
+  struct ArrowArrayStream arrowStream;
+  exportArrowStream(reader, &arrowStream);
+  auto plan = std::make_shared<core::ArrowStreamNode>(
+      "0", type, std::make_shared<ArrowArrayStream>(arrowStream));
+  assertQuery(plan, "SELECT * FROM tmp");
+}
+
+TEST_F(ArrowStreamTest, error) {
+  vector_size_t size = 1'000;
+  std::vector<RowVectorPtr> vectors;
+  for (int32_t i = 0; i < 3; ++i) {
+    vectors.push_back(makeRowVector({makeFlatVector<int32_t>(
+        size, [](auto row) { return row; }, nullEvery(5))}));
+  }
+  createDuckDbTable(vectors);
+  auto type = asRowType(vectors[0]->type());
+
+  // Assert getNext error message.
+  auto reader =
+      std::make_shared<ArrowReader>(pool_, vectors, type, true, false);
+  struct ArrowArrayStream arrowStream;
+  exportArrowStream(reader, &arrowStream);
+  auto plan = std::make_shared<core::ArrowStreamNode>(
+      "0", type, std::make_shared<ArrowArrayStream>(arrowStream));
+  struct ArrowArray arrowArray;
```

Comment:
Let's make sure execution the query plan fails with the right error message:

```
VELOX_ASSERT_THROW(
      AssertQueryBuilder(plan).copyResults(pool()), "Getting next failed");
```

Reply:
Hi Masha, thanks for your advice. I tried that but found a segmentation fault was triggered after throwing a Velox exception in `ArrowStream::getOutput()`. So the exception cannot be caught, and the log is as below.

```
E0209 13:02:29.064574 2900527 Exceptions.h:68] Line: ../../velox/exec/ArrowStream.cpp:42, Function:getOutput, Expression:  Failed to call get_next on ArrowStream: Getting next failed., Source: RUNTIME, ErrorCode: INVALID_STATE
*** Aborted at 1675947749 (Unix time, try 'date -d @1675947749') ***
*** Signal 11 (SIGSEGV) (0x0) received by PID 2900415 (pthread TID 0x7f24e0428700) (linux TID 2900527) (code: address not mapped to object), stack trace: ***
(error retrieving stack trace)
Segmentation fault (core dumped)
```
I assume it is because this exception causes crash for the subsequent computing. Do you have any advice for me?

Reply:
@rui-mo We need to investigate and fix. Errors should not result in crashes.

Here, I think this is because ArrowStream::close is called twice. The guard in the destruction only blocks the call if operator finished successfully, but doesn't work in case of failure. I'd change 'close' to something like this and remove the if-finished check from the destructor.

```
void ArrowStream::close() {
  if (arrowStream_->release) {
     arrowStream_->release(arrowStream_.get());
  }
  SourceOperator::close();
}
```

Reply:
Thank you Masha. This is very useful to me. Through debugging I found the close method was called twice indeed, and I made some improvement according to your suggestion.